refactor(frontend): Providers as hooks

luixo opened the pull request:
That's a big one!
Moving to hooks and straightforward WAMP API getting things more readable (from my perspective) and reusable.
The API itself still needs some revamp, but it should be definitely done after backend typescriptization.
Feel free to criticize this PR, as sometimes I definitely got pulled into refactoring per se :)

luixo commented on frontend/src/pages/transactions/[hash].tsx:
```diff
@@ -98,18 +104,127 @@ const TransactionDetailsPage: NextPage<Props> = (props) => {
   const hash = params!.hash;
   try {
+    const nearNetwork = getNearNetwork(req);
+    const transactionBaseInfo = await wampApi.call(
```
This convertation is a big mess, we need to find out the way to make it more clear.

luixo reviewed (commented):
Please review this PR real good, I started touching fragile things now!

luixo commented on frontend/src/pages/accounts/[id].tsx:
```diff
@@ -99,10 +100,21 @@ export const getServerSideProps: GetServerSideProps<
     if (isAccountExist) {
       try {
-        const account = await new AccountsApi(req).getAccountInfo(id);
+        const account = await Promise.all([
```
Upd: made `providers` folder with couple of those

frol commented on frontend/src/hooks/data/contracts.ts:
```diff
@@ -0,0 +1,44 @@
+        // codeHash does not exist for deleted accounts
+        const account = await wampCall("nearcore-view-account", [accountId]);
+        const codeHash = account["code_hash"];
+        // TODO: find out why we have magic number here
```
If account exists it always has codeHash. That is due to the fact that each account on NEAR can have a contract deployed to it (one or none contract can be there), and when there is some contract deployed, it will have 32 bytes of sha256 hash, otherwise, if no contract has ever been deployed to an account, it will be 32 zero bytes which are encoded as 32 "1" in base58 encoding.

luixo replied:
Thanks!
The constant probably should be called `NO_CONTRACT_FLAG` or something like that.
I'll leave a link to your comment for now and will fix it later.

frol commented:
Somehow, Account Details page does not work (always gives "does not exist"):

https://near-explorer-experimental-frontend-with-swt6.onrender.com/accounts/frol.near

frol commented on frontend/src/hooks/use-wamp-data.ts:
```diff
@@ -0,0 +1,23 @@
+  );
+  useEffect(() => {
+    void fetchValue().catch((error) => {
+      console.error(`${hookName} failed to fetch data due to:`);
```
Yes, we definitely want to integrate Sentry or something like that to catch the errors earlier

luixo commented on frontend/src/libraries/wamp/connection.ts:
```diff
@@ -0,0 +1,76 @@
+      connection.onopen = (session) => {
+        resolve(session);
+      };
+      connection.onclose = (reason) => {
```
We'll track it [here](https://github.com/near/near-explorer/issues/861).

luixo commented:
> Somehow, Account Details page does not work (always gives "does not exist"):

It is actually a good question why.
I found out server-side sends us 'session not open' error (probably [this](https://github.com/crossbario/autobahn-js/blob/bd250753432a7bc61b7b7dae1df5fc70ddedcad6/packages/autobahn/lib/session.js#L1215) one).
Do you have any ideas how this could happen? I bet that on server restart the session will keep up and running again (but let's try it first). Could it just be expired somehow?

luixo commented:
> I bet that on server restart the session will keep up and running again (but let's try it first)

I lost my bet. No good ideas at the moment.

frol commented:
> Do you have any ideas how this could happen? I bet that on server restart the session will keep up and running again (but let's try it first). Could it just be expired somehow?

I don't have any wild guesses

luixo commented:
What I found out is that running server fresh is opening account page all good so far.

frol commented:
> What I found out is that running server fresh is opening account page all good so far.

What do you mean? My account still gives me "account does not exist"

luixo commented:
> What do you mean? My account still gives me "account does not exist"

Yesterday it was okay, today it broke again. I'll investigate.

luixo commented:
It seems like I [overlooked](https://github.com/near/near-explorer/blob/master/frontend/src/libraries/explorer-wamp/index.ts#L67) `isOpen` condition in a current version of connection.
Let's test the current version for a day or two to give it a chance to fail, and I'm ready to merge if it's all good :)

shelegdmitriy commented:
@luixo pages are working fine but it seems something wrong on blocks page (or with ListHandler). Here is the same block (non-existent block) on production
<img width="1792" alt="Снимок экрана 2022-01-26 в 12 16 41" src="https://user-images.githubusercontent.com/34593263/151145247-da3ece6a-a056-42da-b91a-c15d5a2a26f3.png">
and on preview link
<img width="1792" alt="Снимок экрана 2022-01-26 в 12 16 45" src="https://user-images.githubusercontent.com/34593263/151145277-9504f2d2-ce98-40c5-b206-ff8228ada782.png">

luixo commented:
Fixed the one with `blocks/[hash]` page.
It still breaks, now it seems that `connection` reuse is not as good, I simplified it to creating a new connection for every session.
If you have a good idea why do we need both abstractions - connection and session - please let me know

frol commented on frontend/src/libraries/wamp/types.ts:
```diff
@@ -0,0 +1,579 @@
+  | "newcomer"
+  | "on-hold";
+
+export interface BaseValidationNodeInfo {
+  account_id: string;
+  is_slashed?: boolean;
+  num_produced_blocks?: number;
+  num_expected_blocks?: number;
+  public_key?: string;
+  currentStake: string;
+  proposedStake?: string;
+  cumulativeStakeAmount?: string;
+  stakingStatus?: StakingStatus;
+  networkHolder?: boolean;
+  shards?: [number];
+  nodeInfo?: NodeInfo;
+}
```
At some point we will need to make a decision about the naming and enforce it. I prefer to use snake case when passing data over the wire and then client code should just map it over to the style convention on the receiving side, but it is cumbersome and we can argue that Explorer backend is just an extension of Explorer frontend and RPC being a procedure call can follow a single convention of the code-base.

@luixo I feel that at this point whichever strategy is better than this mix of snake-case and camel-case fields floating around.

luixo replied:
In this PR I tried to make as little excessive modifications as I could, and changing casing is one of those.
At some point I'll introduce linting with casing rules so we'll never have to argue on that anymore :)

frol reviewed (approved):
@luixo Thanks for bringing this great sanity improvement to the data handling!

luixo commented on frontend/src/libraries/wamp/connection.ts:
```diff
@@ -0,0 +1,100 @@
+  topic: T,
+  handler: (data: SubscriptionTopicTypes[T]) => void
+): Promise<void> => {
+  if (wampSubscriptionCache[topic]) {
+    return;
+  }
```
Yes, in this implementation we keep one (umbrella) handler for a subscription which invokes an arrays of actual handlers, that are subscribed in `wamp/api`.
Is there a reason we actually need more than one subscriber to a particular topic?

frol replied:
My bad, I missed the implementation in api.ts file.

luixo commented on frontend/src/libraries/wamp/connection.ts:
```diff
@@ -0,0 +1,100 @@
+      max_retries: Number.MAX_SAFE_INTEGER,
+      max_retry_delay: 10,
+    });
+    connection.onopen = (session) => {
+      resolve(session);
+    };
```
Yep, we will handle reconnections separately as it requires extensive testing and out of the scope of this PR.
